Add service worker InstallEvent skeleton

youennf opened the pull request:
#### 041ed08c270ce08b5b96c1f344e8718bc79a091d
<pre>
Add service worker InstallEvent skeleton
<a href="https://rdar.apple.com/144076933">rdar://144076933</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=286911">https://bugs.webkit.org/show_bug.cgi?id=286911</a>

Reviewed by Brady Eidson.

We add a feature flag and expose InstallEvent based on it.
When flag is off, we use ExtendableEvent.

We do not implement yet InstallEvent::addRoutes, this will be for a future patch.

* LayoutTests/imported/w3c/web-platform-tests/service-workers/idlharness.https.any.serviceworker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::URLPattern::create):
* Source/WebCore/Modules/url-pattern/URLPattern.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/workers/service/InstallEvent.cpp: Added.
(WebCore::InstallEvent::InstallEvent):
(WebCore::InstallEvent::addRoutes):
* Source/WebCore/workers/service/InstallEvent.h: Added.
* Source/WebCore/workers/service/InstallEvent.idl: Added.
* Source/WebCore/workers/service/RouterCondition.h: Added.
(WebCore::RouterNotCondition::value const):
(WebCore::RouterNotCondition::value):
(WebCore::RouterNotCondition::RouterNotCondition):
* Source/WebCore/workers/service/RouterCondition.idl: Added.
* Source/WebCore/workers/service/RouterRule.h: Added.
* Source/WebCore/workers/service/RouterRule.idl: Added.
* Source/WebCore/workers/service/RouterSourceDict.h: Added.
(WebCore::RouterSourceDict::isolatedCopy):
* Source/WebCore/workers/service/RouterSourceDict.idl: Added.
* Source/WebCore/workers/service/RouterSourceEnum.h: Added.
* Source/WebCore/workers/service/RouterSourceEnum.idl: Added.
* Source/WebCore/workers/service/RunningStatus.h: Added.
* Source/WebCore/workers/service/RunningStatus.idl: Added.
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::createInstallEvent):
(WebCore::ServiceWorkerThread::queueTaskToFireInstallEvent):

Canonical link: <a href="https://commits.webkit.org/289921@main">https://commits.webkit.org/289921@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7287f7c096cdee332db6f53887974ad3a8f0fac0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88431 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42867 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93389 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39185 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90482 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8337 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16134 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68207 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25920 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91433 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/6373 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/79986 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48573 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/87930 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6145 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34399 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38293 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/81230 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76519 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35285 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95231 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/87207 "Found 1 new JSC binary failure: testapi (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15606 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/88016 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15862 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76330 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18781 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20715 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19135 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8643 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15622 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/109700 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15363 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26377 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18812 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17145 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -6688,6 +6688,20 @@ ServiceWorkerEntitlementDisabledForTesting:
+  status: testable
+  category: networking
+  humanReadableName: "Service Worker Install Event"
+  humanReadableDescription: "Enable Service Worker  Install Event API"
```
Extra space

youennf replied:
Will fix.

beidson commented on Source/WebCore/workers/service/RouterCondition.h:
```diff
@@ -0,0 +1,69 @@
+namespace WebCore {
+
+struct RouterCondition;
+class RouterNotCondition {
```
I'm confused about this class - it has a weird name and is seemingly never used.

youennf replied:
It is used within RouterCondition.
I introduced it to break the potential infinite loop of `struct A { std::optional<A> a ; }`
RouterNotCondition takes a RouterCondition so that the binding generated code works fine.

beidson reviewed (approved):
r+ with a nit and a question.

youennf committed:
Add service worker InstallEvent skeleton
rdar://144076933
https://bugs.webkit.org/show_bug.cgi?id=286911

Reviewed by Brady Eidson.

We add a feature flag and expose InstallEvent based on it.
When flag is off, we use ExtendableEvent.

We do not implement yet InstallEvent::addRoutes, this will be for a future patch.

* LayoutTests/imported/w3c/web-platform-tests/service-workers/idlharness.https.any.serviceworker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::URLPattern::create):
* Source/WebCore/Modules/url-pattern/URLPattern.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/workers/service/InstallEvent.cpp: Added.
(WebCore::InstallEvent::InstallEvent):
(WebCore::InstallEvent::addRoutes):
* Source/WebCore/workers/service/InstallEvent.h: Added.
* Source/WebCore/workers/service/InstallEvent.idl: Added.
* Source/WebCore/workers/service/RouterCondition.h: Added.
(WebCore::RouterNotCondition::value const):
(WebCore::RouterNotCondition::value):
(WebCore::RouterNotCondition::RouterNotCondition):
* Source/WebCore/workers/service/RouterCondition.idl: Added.
* Source/WebCore/workers/service/RouterRule.h: Added.
* Source/WebCore/workers/service/RouterRule.idl: Added.
* Source/WebCore/workers/service/RouterSourceDict.h: Added.
(WebCore::RouterSourceDict::isolatedCopy):
* Source/WebCore/workers/service/RouterSourceDict.idl: Added.
* Source/WebCore/workers/service/RouterSourceEnum.h: Added.
* Source/WebCore/workers/service/RouterSourceEnum.idl: Added.
* Source/WebCore/workers/service/RunningStatus.h: Added.
* Source/WebCore/workers/service/RunningStatus.idl: Added.
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::createInstallEvent):
(WebCore::ServiceWorkerThread::queueTaskToFireInstallEvent):

Canonical link: https://commits.webkit.org/289921@main
